packages/kbn-custom-events/src/index.ts

```diff
@@ -0,0 +1,180 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { PackageInfo } from '@kbn/config';
+import { isArray, isInteger, mapKeys, snakeCase } from 'lodash';
+import { FullStoryApi } from './fullstory';
+
+export interface CustomEventConfig {
+  enabled: boolean;
+  fullstoryOrgId?: string;
+  esOrgId?: string;
+  basePath: any;
+  userIdPromise: Promise<string | undefined>;
+  packageInfo: PackageInfo;
+}
+
+// based on https://help.fullstory.com/hc/en-us/articles/360020623234#Custom%20Event%20Name%20Requirements
+type FULLSTORY_EVENT_TYPES = string | number | boolean | Date;
+type FullstoryEventValue = FULLSTORY_EVENT_TYPES | FULLSTORY_EVENT_TYPES[] | undefined;
+
+class CustomEvents {
```

Comment:
@afharo have you had a chance to review the proposed implementation for event-based telemetry?

Reply:
Done! Thanks for the ping! 🧡